File: ios/Packages/Domain/Sources/Domain/Combiner.swift
// Deterministic rule-based combiner (PLAN §5.6): the last resort after the local and Gemini planners, and the "reused" tier. Mirrors combiner.ts.
import Foundation

public struct Combiner: Sendable {
    private let taxonomy: Taxonomy
    private let rules: TemperatureRules
    private let validator: OutfitValidator

    public init(taxonomy: Taxonomy = .shared, rules: TemperatureRules = .shared, validator: OutfitValidator = OutfitValidator()) {
        self.taxonomy = taxonomy; self.rules = rules; self.validator = validator
    }

    private func sorted(_ cands: [Candidate], _ category: Category) -> [Candidate] {
        cands.filter { $0.category == category }.sorted { $0.score > $1.score }
    }

    private func reason(_ slot: Slot, _ c: Candidate, _ ctx: PlanContext) -> String {
        let w = ctx.wearWindow
        let colour = c.colorName.replacingOccurrences(of: "_", with: " ")
        let sub = c.subcategory.replacingOccurrences(of: "_", with: " ")
        switch slot {
        case .outerwear: return "\(c.warmth.rawValue) layer for \(Int(w.minFeelsLikeC.rounded())) °C"
        case .midLayer: return "mid layer for a \(rules.band(w.minFeelsLikeC)) start"
        case .shoes: return "\(colour) \(sub) for \(ctx.occasion.rawValue)"
        default: return "\(colour) \(sub), \(ctx.occasion.rawValue)-ready"
        }
    }

    private func build(core: [(Slot, Candidate)], cands: [Candidate], ctx: PlanContext) -> PlannedOutfit? {
        let w = ctx.wearWindow
        var slots = core.map { PlannedSlot(slot: $0.0, itemId: $0.1.id, reason: reason($0.0, $0.1, ctx)) }
        var used = Set(core.map { $0.1.id })
        func add(_ slot: Slot, _ item: Candidate?) {
            guard let item, !used.contains(item.id) else { return }
            used.insert(item.id)
            slots.append(PlannedSlot(slot: slot, itemId: item.id, reason: reason(slot, item, ctx)))
        }

        add(.shoes, sorted(cands, .shoes).first)
        if rules.outerwearAllowed(w, occasion: ctx.occasion) {
            let outer = sorted(cands, .outerwear).first { !rules.outerwearForbidden(w, itemWarmth: $0.warmth) }
            if let outer, rules.outerwearRequired(w) || w.minFeelsLikeC < 15 { add(.outerwear, outer) }
        }
        if rules.midLayerAllowed(w) && w.minFeelsLikeC < 15 && !slots.contains(where: { $0.slot == .outerwear }) { add(.midLayer, sorted(cands, .midLayer).first) }

        let limits = taxonomy.limits(for: ctx.accessoryCount)
        add(.bag, sorted(cands, .bag).first)
        var jewelSubs = Set<String>()
        for j in sorted(cands, .jewelry) {
            if jewelSubs.count >= min(limits.jewelry, 2) { break }
            if jewelSubs.contains(j.subcategory) { continue }
            jewelSubs.insert(j.subcategory)
            add(.jewelry, j)
        }
        if limits.accessory > 0 { add(.accessory, sorted(cands, .accessory).first) }

        let byId = Dictionary(uniqueKeysWithValues: cands.map { ($0.id, $0) })
        var colours: [String] = []
        for s in slots { if let c = byId[s.itemId] { let n = c.colorName.replacingOccurrences(of: "_", with: " "); if !colours.contains(n) { colours.append(n) } } }
        let minT = Int(w.minFeelsLikeC.rounded()), maxT = Int(w.maxFeelsLikeC.rounded())
        var outfit = PlannedOutfit(
            slots: slots,
            rationale: "Rule-based pick: \(colours.prefix(3).joined(separator: ", ")) for \(ctx.occasion.rawValue), \(minT)–\(maxT) °C.",
            weatherFit: .acceptable,
            formality: core.first?.1.formality ?? .casual,
            palette: colours,
            layeringNote: rules.spansTwoBands(w) ? "\(minT)° early, \(maxT)° later: shed the outer layer as it warms." : nil,
            confidence: 0.5)

        var result = validator.validate(outfit, candidates: byId, context: ctx)
        for slot in [Slot.accessory, .jewelry, .bag, .midLayer, .outerwear] where !result.passed {
            guard result.rulesFailed.contains(where: { $0.contains(slot.rawValue) }) else { continue }
            outfit.slots.removeAll { $0.slot == slot }
            result = validator.validate(outfit, candidates: byId, context: ctx)
        }
        return result.passed ? outfit : nil
    }

    /// Try one-piece and top+bottom cores in score order; return the first outfit the validator accepts.
    public func combine(_ cands: [Candidate], ctx: PlanContext) -> PlannedOutfit? {
        let anchor = ctx.anchorId.flatMap { id in cands.first { $0.id == id } }
        let tops = sorted(cands, .top), bottoms = sorted(cands, .bottom), onePieces = sorted(cands, .onePiece)

        var cores: [[(Slot, Candidate)]] = []
        func push(_ core: [(Slot, Candidate)]) {
            if let anchor, [Category.top, .bottom, .onePiece].contains(anchor.category), !core.contains(where: { $0.1.id == anchor.id }) { return }
            cores.append(core)
        }
        for op in onePieces { push([(.onePiece, op)]) }
        for t in tops.prefix(4) { for b in bottoms.prefix(4) { push([(.top, t), (.bottom, b)]) } }

        var pool = cands
        if let anchor {
            var boosted = anchor
            boosted.score = .greatestFiniteMagnitude
            pool = [boosted] + cands.filter { $0.id != anchor.id }
        }
        for core in cores {
            if let outfit = build(core: core, cands: pool, ctx: ctx) { return outfit }
        }
        return nil
    }
}

File: backend/src/domain/combiner.ts
// Deterministic rule-based combiner (PLAN.md §5.6: the fallback after a failed repair, and the "reused" daily tier). The card always renders.
import { accessoryLimits, type Slot } from "./taxonomy.js";
import { band, midLayerAllowed, outerwearAllowed, outerwearRequired, spansTwoBands } from "./temperature.js";
import { validateOutfit } from "./validator.js";
import type { Candidate, PlanContext, PlannedOutfit, PlannedSlot } from "./types.js";

function sorted(cands: Candidate[], category: string): Candidate[] {
  return cands.filter((c) => c.category === category).sort((a, b) => b.score - a.score);
}

function reason(slot: Slot, c: Candidate, ctx: PlanContext): string {
  const w = ctx.wearWindow;
  switch (slot) {
    case "outerwear":
      return `${c.warmth} layer for ${Math.round(w.min_feels_like_c)} °C`;
    case "mid_layer":
      return `mid layer for a ${band(w.min_feels_like_c)} start`;
    case "shoes":
      return `${c.color_name.replace(/_/g, " ")} ${c.subcategory.replace(/_/g, " ")} for ${ctx.occasion}`;
    default:
      return `${c.color_name.replace(/_/g, " ")} ${c.subcategory.replace(/_/g, " ")}, ${ctx.occasion}-ready`;
  }
}

function build(core: { slot: Slot; item: Candidate }[], cands: Candidate[], ctx: PlanContext): PlannedOutfit | null {
  const w = ctx.wearWindow;
  const slots: PlannedSlot[] = core.map(({ slot, item }) => ({ slot, item_id: item.id, reason: reason(slot, item, ctx) }));
  const used = new Set(core.map((c) => c.item.id));
  const add = (slot: Slot, item: Candidate | undefined) => {
    if (!item || used.has(item.id)) return;
    used.add(item.id);
    slots.push({ slot, item_id: item.id, reason: reason(slot, item, ctx) });
  };

  const shoes = sorted(cands, "shoes");
  add("shoes", shoes[0]);

  if (outerwearAllowed(w, ctx.occasion)) {
    const outer = sorted(cands, "outerwear").find((o) => !(w.max_feels_like_c >= 24 && o.warmth !== "light"));
    if (outer && (outerwearRequired(w) || w.min_feels_like_c < 15)) add("outerwear", outer);
  }
  if (midLayerAllowed(w) && w.min_feels_like_c < 15 && !slots.some((s) => s.slot === "outerwear")) add("mid_layer", sorted(cands, "mid_layer")[0]);

  const limits = accessoryLimits(ctx.accessoryCount);
  add("bag", sorted(cands, "bag")[0]);
  const jewelry = sorted(cands, "jewelry");
  const jewelSubs = new Set<string>();
  for (const j of jewelry) {
    if (jewelSubs.size >= Math.min(limits.jewelry, 2)) break;
    if (jewelSubs.has(j.subcategory)) continue;
    jewelSubs.add(j.subcategory);
    add("jewelry", j);
  }
  if (limits.accessory > 0) add("accessory", sorted(cands, "accessory")[0]);

  const colours = [...new Set(slots.map((s) => cands.find((c) => c.id === s.item_id)?.color_name.replace(/_/g, " ")).filter(Boolean))];
  const layering = spansTwoBands(w)
    ? `${Math.round(w.min_feels_like_c)}° early, ${Math.round(w.max_feels_like_c)}° later: shed the outer layer as it warms.`
    : null;
  const outfit: PlannedOutfit = {
    slots,
    rationale: `Rule-based pick: ${colours.slice(0, 3).join(", ")} for ${ctx.occasion}, ${Math.round(w.min_feels_like_c)}–${Math.round(w.max_feels_like_c)} °C.`,
    weather_fit: "acceptable",
    formality: core[0]?.item.formality ?? "casual",
    palette: colours.map(String),
    layering_note: layering,
    confidence: 0.5,
  };

  // Drop optional slots one by one until the validator passes.
  const optional: Slot[] = ["accessory", "jewelry", "bag", "mid_layer", "outerwear"];
  const byId = new Map(cands.map((c) => [c.id, c]));
  let result = validateOutfit(outfit, byId, ctx);
  for (const slot of optional) {
    if (result.passed) break;
    if (!result.rules_failed.some((r) => r.includes(slot))) continue;
    outfit.slots = outfit.slots.filter((s) => s.slot !== slot);
    result = validateOutfit(outfit, byId, ctx);
  }
  return result.passed ? outfit : null;
}

/** Try one-piece and top+bottom cores in score order; return the first outfit the validator accepts. */
export function combine(cands: Candidate[], ctx: PlanContext): PlannedOutfit | null {
  const anchor = ctx.anchorId ? cands.find((c) => c.id === ctx.anchorId) : undefined;
  const tops = sorted(cands, "top");
  const bottoms = sorted(cands, "bottom");
  const onePieces = sorted(cands, "one_piece");

  const cores: { slot: Slot; item: Candidate }[][] = [];
  const pushIfAnchorOk = (core: { slot: Slot; item: Candidate }[]) => {
    if (anchor && !core.some((c) => c.item.id === anchor.id) && ["top", "bottom", "one_piece"].includes(anchor.category)) return;
    cores.push(core);
  };
  for (const op of onePieces) pushIfAnchorOk([{ slot: "one_piece", item: op }]);
  for (const t of tops.slice(0, 4)) for (const b of bottoms.slice(0, 4)) pushIfAnchorOk([{ slot: "top", item: t }, { slot: "bottom", item: b }]);

  // Anchor in a non-core slot: force it into the candidate pool front so `add` picks it first.
  const pool = anchor ? [anchor, ...cands.filter((c) => c.id !== anchor.id)].map((c, i) => (i === 0 ? { ...c, score: Number.MAX_SAFE_INTEGER } : c)) : cands;

  for (const core of cores) {
    const outfit = build(core, pool, ctx);
    if (outfit) return outfit;
  }
  return null;
}
